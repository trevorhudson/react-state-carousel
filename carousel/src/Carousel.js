import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";


/** Carousel: displays images and arrows to navigate through them
 *
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 *
 * State:
 * - currCardIdx: integer for current card index
 *
 * App --> Carousel --> Card
 */
function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);

  console.log('Current Card', currCardIdx, setCurrCardIdx);
  const currCard = photos[currCardIdx];
  const total = photos.length;

  //Increments currCardIdx state by 1 -> loops at end.
  function goForward() {
    setCurrCardIdx(currCardIdx === total - 1 ? 0 : currCardIdx + 1);
  }

  //Decrements currCardIdx state by 1 -> loops at end.
  function goBackward() {
    setCurrCardIdx(currCardIdx === 0 ? total - 1 : currCardIdx - 1);
  }

  return (
    <div className="Carousel">
      <h1>{title}</h1>
      <div className="Carousel-main">

        <i
          className="bi bi-arrow-left-circle"
          onClick={goBackward}
          style={{ "visibility": currCardIdx === 0 ? "hidden" : "visible" }}
        />

        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        <i
          className="bi bi-arrow-right-circle"
          onClick={goForward}
          style={{ "visibility": currCardIdx === 2 ? "hidden" : "visible" }}

        />
      </div>
    </div>
  );
}

export default Carousel;
