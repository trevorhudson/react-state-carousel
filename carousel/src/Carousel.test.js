import { render, fireEvent, debug } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";

//snapshot tests -> for individual actions
describe("Carousel tests", function () {

  it("renders the Carousel without crashing", function () {

    render(< Carousel photos={TEST_IMAGES} title={"Test carouselTitle"} />);
  });

  it("matches the snapshot", function () {
    const { container } = render(
      < Carousel photos={TEST_IMAGES} title={"Test carouselTitle"} />);

    expect(container).toMatchSnapshot();
  });

  it("works when you click on the right arrow", function () {
    const { container } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );
    // expect the first image to show, but not the second
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 3"]')
    ).not.toBeInTheDocument();

    // move forward in the carousel
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);

    // expect the second image to show, but not the first
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 3"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).toBeInTheDocument();
  });

  it("matches snapshot after clicking right arrow", function () {
    const { container } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);
    expect(container).toMatchSnapshot();
  });

  it("works when you click on the left arrow", function () {
    const { container } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );

    // move forward in the carousel
    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);


    // expect the first image to show, but not the second, or third
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 3"]')
    ).not.toBeInTheDocument();


    // move backwards in the carousel
    const leftArrow = container.querySelector(".bi-arrow-left-circle");
    fireEvent.click(leftArrow);

    // expect the last image to show, but not the first
    expect(
      container.querySelector('img[alt="testing image 2"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 3"]')
    ).not.toBeInTheDocument();
    expect(
      container.querySelector('img[alt="testing image 1"]')
    ).toBeInTheDocument();
  });

  it("matches snapshot after clicking left arrow", function () {
    const { container } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );

    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);
    const leftArrow = container.querySelector(".bi-arrow-left-circle");
    fireEvent.click(leftArrow);

    expect(container).toMatchSnapshot();
  });

  // change name -> first image
  it("left arrow hidden on first image", function () {
    const { container } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );
    const leftArrow = container.querySelector(".bi-arrow-left-circle");
    expect(leftArrow).toBeVisible();

    expect(
      container.querySelector('.bi-arrow-left-circle')
    ).toBeInTheDocument();

    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    fireEvent.click(rightArrow);

    expect(leftArrow).not.toBeVisible();

    expect(
      container.querySelector('.bi-arrow-left-circle')
    ).toBeInTheDocument();

  });

  it("right arrow hidden on last image ", function () {
    const { container } = render(
      <Carousel
        photos={TEST_IMAGES}
        title="images for testing"
      />
    );

    const rightArrow = container.querySelector(".bi-arrow-right-circle");
    expect(
      container.querySelector('.bi-arrow-right-circle')
    ).toBeInTheDocument();

    expect(rightArrow).toBeVisible();

    fireEvent.click(rightArrow);
    fireEvent.click(rightArrow);

    expect(rightArrow).not.toBeVisible();

    expect(
      container.querySelector('.bi-arrow-right-circle')
    ).not.toBeInTheDocument();

  });

})




