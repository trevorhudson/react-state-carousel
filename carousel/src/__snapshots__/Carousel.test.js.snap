// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel tests matches snapshot after clicking left arrow 1`] = `
<div>
  <div
    class="Carousel"
  >
    <h1>
      images for testing
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
        style="visibility: hidden;"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 1
        </h4>
        <img
          alt="testing image 1"
          class="Card-image"
          src="test1.com"
        />
        <small
          class="Card-small"
        >
          Image 
          1
           of 
          3
          .
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        style="visibility: visible;"
      />
    </div>
  </div>
</div>
`;

exports[`Carousel tests matches snapshot after clicking right arrow 1`] = `
<div>
  <div
    class="Carousel"
  >
    <h1>
      images for testing
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
        style="visibility: visible;"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 2
        </h4>
        <img
          alt="testing image 2"
          class="Card-image"
          src="test2.com"
        />
        <small
          class="Card-small"
        >
          Image 
          2
           of 
          3
          .
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        style="visibility: visible;"
      />
    </div>
  </div>
</div>
`;

exports[`Carousel tests matches the snapshot 1`] = `
<div>
  <div
    class="Carousel"
  >
    <h1>
      Test carouselTitle
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
        style="visibility: hidden;"
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 1
        </h4>
        <img
          alt="testing image 1"
          class="Card-image"
          src="test1.com"
        />
        <small
          class="Card-small"
        >
          Image 
          1
           of 
          3
          .
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        style="visibility: visible;"
      />
    </div>
  </div>
</div>
`;
